### PR TITLE
Native arrays, Db Arrays, and types.

### PR DESCRIPTION
**Pull request by devinrsmith**

### CODE REVIEW

**Comment by devinrsmith** [comment]
Documentation needed: wrt examples on constructing new tables explicitly w/ db array values or native arrays as values.

File: qst/src/main/java/io/deephaven/qst/type/NativeArrayType.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+@Immutable
+@SimpleStyle
+public abstract class NativeArrayType<T, ComponentType> extends ArrayTypeBase<T, ComponentType> {
+
+    public static <T, ComponentType> NativeArrayType<T, ComponentType> of(Class<T> arrayType,
+        Type<ComponentType> componentType) {
+        return ImmutableNativeArrayType.of(arrayType, componentType);
+    }
+
+    public static <ComponentType> NativeArrayType<?, ComponentType> toArrayType(
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/NativeArrayType.java:26]: I feel like these methods should be *on* the primitive types without an arg.

File: qst/src/main/java/io/deephaven/qst/type/DbPrimitiveArrayType.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ * @param <ComponentType> the component type
+ */
+@Immutable
+@SimpleStyle
+public abstract class DbPrimitiveArrayType<T, ComponentType>
+    extends ArrayTypeBase<T, ComponentType> {
+
+    public static <T, ComponentType> DbPrimitiveArrayType<T, ComponentType> of(Class<T> clazz,
+        PrimitiveType<ComponentType> primitiveType) {
+        return ImmutableDbPrimitiveArrayType.of(clazz, primitiveType);
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/DbPrimitiveArrayType.java:23]: This is recursive.

File: qst/src/main/java/io/deephaven/qst/type/DbPrimitiveArrayType.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        InvocationTargetException, NoSuchMethodException, IllegalAccessException {
+        return invokeTypeMethod("io.deephaven.db.tables.dbarrays.DbFloatArray");
+    }
+
+    public static DbPrimitiveArrayType<?, Double> doubleType() throws ClassNotFoundException,
+        InvocationTargetException, NoSuchMethodException, IllegalAccessException {
+        return invokeTypeMethod("io.deephaven.db.tables.dbarrays.DbDoubleArray");
+    }
+
+    private static <ComponentType> DbPrimitiveArrayType<?, ComponentType> invokeTypeMethod(
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/DbPrimitiveArrayType.java:66]: We should be sure to have unit tests for all the callers of this method.

File: qst/src/main/java/io/deephaven/qst/type/Type.java [hunk 4]
```diff
@@ -12,19 +13,29 @@
...
+        if (clazz.isArray()) {
+            return NativeArrayType.of(clazz, find(clazz.getComponentType()));
+        }
+        return found.orElseGet(() -> CustomType.of(clazz));
     }
 
     /**
-     * The list of known types. Includes the universe of all {@link PrimitiveType primitive types}
-     * and {@link GenericType generic types}, except for {@link CustomType custom types}.
+     * The list of known types. Includes the universe of {@link PrimitiveType primitive types} and
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/Type.java:35]: There's a stray period, and my brain can't parse the existing formulation, please revisit.

File: qst/src/main/java/io/deephaven/qst/type/Type.java [hunk 5]
```diff
@@ -12,19 +13,29 @@
...
     static <T> Type<T> find(Class<T> clazz) {
-        return TypeHelper.findStatic(clazz).orElseGet(() -> CustomType.of(clazz));
+        Optional<Type<T>> found = TypeHelper.findStatic(clazz);
+        if (found.isPresent()) {
+            return found.get();
+        }
+        if (clazz.isArray()) {
+            return NativeArrayType.of(clazz, find(clazz.getComponentType()));
+        }
+        return found.orElseGet(() -> CustomType.of(clazz));
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/Type.java:31]: You already know `found` is not present. Can just do `CustomType.of`.

File: qst/src/main/java/io/deephaven/qst/type/Type.java [hunk 6]
```diff
@@ -12,19 +13,29 @@
...
     /**
-     * Finds the {@link #knownTypes() known type}, or else creates a {@link CustomType custom type}.
+     * Finds the {@link #knownTypes() known type}, or else creates the relevant
+     * {@link NativeArrayType native array type} or {@link CustomType custom type}.
      *
      * @param clazz the class
      * @param <T> the generic type of {@code clazz}
      * @return the type
      */
     static <T> Type<T> find(Class<T> clazz) {
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/Type.java:23]: Should we break if someone uses this with `DbArray.class`?
Are we happy that this gives weaker typing for native arrays than making the native array type directly?

Comment (devinrsmith) [reply qst/src/main/java/io/deephaven/qst/type/Type.java:23]: Yes - I've added a check against creating Db array custom types.

It's a good question if the user already knows they have an array type, but aren't sure which one. I added `NativeArrayType.find(Class<?>)`.

File: qst/src/main/java/io/deephaven/qst/type/CustomType.java [hunk 7]
```diff
@@ -43,10 +43,17 @@ final void checkClazz() {
...
+    final void checkNotArray() {
         if (clazz().isArray()) {
             throw new IllegalArgumentException(String.format(
                 "Can't create an array type here, use '%s' instead", NativeArrayType.class));
         }
     }
+
+    @Check
+    final void checkNotDbArray() {
+        if (clazz().getName().startsWith("io.deephaven.db.tables.dbarrays.Db")) {
```
Comment (rcaudy) [inline qst/src/main/java/io/deephaven/qst/type/CustomType.java:55]: `!DbArrayBase.class.isAssignableFrom(clazz())` would be a better test.

Comment (devinrsmith) [reply qst/src/main/java/io/deephaven/qst/type/CustomType.java:55]: DbArrayBase is not on the classpath for QST.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**